Plugins/Effect/Effect.cpp

```diff
@@ -0,0 +1,171 @@
+#include "Effect.hpp"
+
+#include "API/Constants.hpp"
+#include "API/Globals.hpp"
+#include "API/CExoString.hpp"
+#include "API/CGameEffect.hpp"
+#include "Utils.hpp"
+#include "ViewPtr.hpp"
+
+#include <string>
+#include <functional>
+
+using namespace NWNXLib;
+
+static ViewPtr<Effect::Effect> g_plugin;
+
+NWNX_PLUGIN_ENTRY Plugin::Info* PluginInfo()
+{
+    return new Plugin::Info
+    {
+        "Effect",
+        "Miscellaneous Effectity functions",
+        "sherincall",
+        "sherincall@gmail.com",
+        1,
+        true
+    };
+}
+
+NWNX_PLUGIN_ENTRY Plugin* PluginLoad(Plugin::CreateParams params)
+{
+    g_plugin = new Effect::Effect(params);
+    return g_plugin;
+}
+
+
+namespace Effect {
+
+Effect::Effect(const Plugin::CreateParams& params)
+    : Plugin(params)
+{
+#define REGISTER(func) \
+    GetServices()->m_events->RegisterEvent(#func, std::bind(&Effect::func, this, std::placeholders::_1))
+
+    REGISTER(PackEffect);
+    REGISTER(UnpackEffect);
+
+#undef REGISTER
+
+}
+
+Effect::~Effect()
+{
+}
+
+ArgumentStack Effect::PackEffect(ArgumentStack&& args)
+{
+    ArgumentStack stack;
+    API::CGameEffect *eff = new API::CGameEffect();
+
+    eff->m_sCustomTag = Services::Events::ExtractArgument<std::string>(args).c_str();
+
+    eff->m_oidParamObjectID[3] = Services::Events::ExtractArgument<API::Types::ObjectID>(args);
+    eff->m_oidParamObjectID[2] = Services::Events::ExtractArgument<API::Types::ObjectID>(args);
+    eff->m_oidParamObjectID[1] = Services::Events::ExtractArgument<API::Types::ObjectID>(args);
+    eff->m_oidParamObjectID[0] = Services::Events::ExtractArgument<API::Types::ObjectID>(args);
+
+    eff->m_sParamString[5] = Services::Events::ExtractArgument<std::string>(args).c_str();
+    eff->m_sParamString[4] = Services::Events::ExtractArgument<std::string>(args).c_str();
+    eff->m_sParamString[3] = Services::Events::ExtractArgument<std::string>(args).c_str();
+    eff->m_sParamString[2] = Services::Events::ExtractArgument<std::string>(args).c_str();
+    eff->m_sParamString[1] = Services::Events::ExtractArgument<std::string>(args).c_str();
+    eff->m_sParamString[0] = Services::Events::ExtractArgument<std::string>(args).c_str();
+
+    eff->m_nParamFloat[3] = Services::Events::ExtractArgument<float>(args);
+    eff->m_nParamFloat[2] = Services::Events::ExtractArgument<float>(args);
+    eff->m_nParamFloat[1] = Services::Events::ExtractArgument<float>(args);
+    eff->m_nParamFloat[0] = Services::Events::ExtractArgument<float>(args);
+
+    eff->SetNumIntegers(8); // allocate array
```

Comment:
I haven't looked at the NWN code for this, but is there any chance that `SetNumIntegers` allocates 8 integers, and then overwriting `m_nNumIntegers` causes the incorrect number of integers to be freed later?

Reply:
They are a continuous array. It'll just do `delete [] m_nParamInteger;`.

Somewhat related, but EffectDamage() actually uses more than 8. But since this is a limitation of a `util` module function and not the core, I'd rather leave at 8 which should cover all other uses, and if someone needs to do EffectDamage() they can elsewhere.